### ISSUE TITLE
Export to Excel using AlaSql 0.4.2 throws error when exporting large data.

### ISSUE THREAD

**Issue by sarveshascbe** [agershun/alasql#945]
I am using AlaSQL v0.4.2, majorly for export JavaScript array data to Excel (XLS) format.

In a JavaScript code object there are approximately 10,000 + records and 88 columns.

I am using export to excel (XLS) option. 

var excelObj = alasql('SELECT * INTO XLS("CounterCorrection.xls",?) FROM ?', [mystyle, data]);

First 1 or 2 times excel data is exported successfully, subsequent attempts to export for same or less volume of data I get errors "Not enough memory" or "Not enough storage available to complete this operation".

This happens in IE9, IE10, IE11 and all other browsers too.

Attached is a sample file generated from alasql, the javascript object contains exactly the same columns and rows as depicted in excel. Headers are columns, and others are rows.
[CounterCorrectionRevised.zip](https://github.com/agershun/alasql/files/1343045/CounterCorrectionRevised.zip)

**Comment by mathiasrw** [comment]
Would be interesting to look into if Alasql has any memory leaks

**Comment by sarveshascbe** [comment]
Is there any fix for these memory leaks ? If so would be great to know about expected ETA.

**Comment by mathiasrw** [comment]
Any pull requests fixing any memory leaks would be much appreciated...

**Comment by sarveshascbe** [comment]
I have the latest pull request code. If there are no memory leaks then why does it not work consistently. It fails for large data sets, and it fails for smaller data sets too on the third or fourth time. Could you please validate my scenario and let know about its solution. If you try to export more than 2 or 3 times for same or more data you will notice that you get out of memory exception or not enough storage available for this operation exception.

**Comment by sarveshascbe** [comment]
Actually surprisingly working fine in Chrome without any issues. But the issues still persist in all IE versions especially if data is big and fails second or third time, after which it stops working to export even small amount of data. Can you please check and let me know.

**Comment by mathiasrw** [comment]
Do you have any test data? How much? how big?

**Comment by sarveshascbe** [comment]
var mystyle = {
                    headers: true,
                    sheetid: 'Counter Correction',
                };

                //Create XLS format using alasql.js file.
                var excelObj = alasql('SELECT * INTO XLS("CounterCorrection.xls",?) FROM ?', [mystyle, data]);
                console.log("Counter Correction Data Export Success");

data - JavaScript array object contains 8881 rows and about 88 columns. In UI on a button click above code is executed.

The exported excel is attached in the zip file "CounterCorrectionAlaSql.zip"

In Firefox and Chrome i don't have any issues, even if i try to export multiple time on click of a button. Whereas in all IE versions i face an issue of either "Not enough storage available for this operation" or "Not enough memory" exception. Once i get this exception even for smaller volume of data, say for example if the data Javascript object contains only 1000 rows, even then it fails. Only way to recover is to restart the browser.
[CounterCorrectionAlaSql.zip](https://github.com/agershun/alasql/files/1361570/CounterCorrectionAlaSql.zip)

**Comment by mathiasrw** [comment]
I cant replicate. Can you please provide full example code?

Is it same result if you use `XLSX`? What if you export to `CSV`?

**Comment by sarveshascbe** [comment]
Export to CSV is working fine. Same error for XLSX too. Basically this happens only for 3rd or 4th time execution, sometimes fails on 2nd time itself.

As per the attached code file "SampleCode.txt". exportCounterCorrectionData is the function being called when click of a button. I tested the code by clicking the button after exporting the file for each step.

What happens when function exportCounterCorrectionData  is being executed it calls another function called ExecutePromises which in turn executes multiple promises by calling a Service REST API - api/SorList/GetCounterCorrectionExportData. Approximately there will be 6 to 8 promises executing in parallel.

Followed by executing the promises , there is a function called ExecuteDataExport where $q will wait for all promises to finish. The complete data retrieved from service call is stored in $scope.counterCorrectionExportData, approximately there will be 6,50,000 objects.

Followed by i order the above scope object by datetimestamp using alasql as follows : var orderedTimeStamps = alasql('SELECT * FROM ? ORDER BY DataTimeStamp ASC', [$scope.counterCorrectionExportData]);

Subsequently i construct new JavaScript object called counterCorrectionData which approximately has about 8881 records, each having 88 columns. This is what is being directly exported using AlaSql, which you observe in excel sheet.

ExportToExcel(counterCorrectionData);

var mystyle = {
                    headers: true,
                    sheetid: 'Counter Correction',
                };

                //Create XLS format using alasql.js file.
                var excelObj = alasql('SELECT * INTO XLS("CounterCorrection.xls",?) FROM ?', [mystyle, data]);
                console.log("Counter Correction Data Export Success");

Object data contains 8881 rows and 88 columns as described earlier.

Now this whole process as explained and shared, works inconsistently, IE throws error on 2nd or 3rd time.

Kindly request you to check on this and let know as soon as possible.

[SampleCode.txt](https://github.com/agershun/alasql/files/1367559/SampleCode.txt)

**Comment by mathiasrw** [comment]
Please provide example code that only illustrates the error. Elements like `var periodStartDate = $scope.parent.hsStartDate;` makes sense to you but does not help the debugging process.

**Comment by sarveshascbe** [comment]
StartDate and EndDate will be 6 months duration. Any case you dont have test data, since data is loaded from our internal rest service, what you can do is read data from the previous attached excel , and convert it javascript array of objects, and then try to export to excel multiple times in Internet Explorer.

**Comment by mathiasrw** [comment]
Please provide a complete test code. Alternatively, you might be able to do a pay-per-hour agreement with @agershun

**Comment by sarveshascbe** [comment]
Error occurs when exporting to XLSX too on 2nd or 3rd time when trying in IE. What is pay per agreement ?

As i said before, you can read the data from my previous attached excel, and store it in a JS array of objects named data. Now from this data JS object try to export to XLS in IE browser many times, for example : every time during a button click. 

So on the button click , you should read from the excel , store it in "data" JS object , where each row in excel is a row in "data" JS object, with columns matching in excel sheet. Then using "data" object export to XLS using IE several times. You will notice the error.

**Comment by sarveshascbe** [comment]
[HistoricalDataController.zip](https://github.com/agershun/alasql/files/1397970/HistoricalDataController.zip)
Please find attached code.

**Comment by SathishRamV** [comment]
Hello , Any New Update on this request. @mathiasrw  , it's takes long time to download 50K row item to CSV.

**Comment by mathiasrw** [comment]
@SathishRamV Please make a new issue and add sample code and data...
